client/web/src/enterprise/batches/detail/changesets/PublishChangesetsModal.tsx

```diff
@@ -0,0 +1,94 @@
+import Dialog from '@reach/dialog'
+import React, { useCallback, useState } from 'react'
+
+import { Form } from '@sourcegraph/branded/src/components/Form'
+import { LoadingSpinner } from '@sourcegraph/react-loading-spinner'
+import { asError, isErrorLike } from '@sourcegraph/shared/src/util/errors'
+
+import { ErrorAlert } from '../../../../components/alerts'
+import { Scalars } from '../../../../graphql-operations'
+import { publishChangesets as _publishChangesets } from '../backend'
+
+export interface PublishChangesetsModalProps {
+    onCancel: () => void
+    afterCreate: () => void
+    batchChangeID: Scalars['ID']
+    changesetIDs: () => Promise<Scalars['ID'][]>
+
+    /** For testing only. */
+    publishChangesets?: typeof _publishChangesets
+}
+
+export const PublishChangesetsModal: React.FunctionComponent<PublishChangesetsModalProps> = ({
+    onCancel,
+    afterCreate,
+    batchChangeID,
+    changesetIDs,
+    publishChangesets: publishChangesets = _publishChangesets,
+}) => {
+    const [isLoading, setIsLoading] = useState<boolean | Error>(false)
+    const [draft, setDraft] = useState(false)
+
+    const onSubmit = useCallback<React.FormEventHandler>(async () => {
+        setIsLoading(true)
+        try {
+            const ids = await changesetIDs()
+            await publishChangesets(batchChangeID, ids, draft)
+            afterCreate()
+        } catch (error) {
+            setIsLoading(asError(error))
+        }
+    }, [changesetIDs, publishChangesets, batchChangeID, draft, afterCreate])
+
+    const onToggleDraft = useCallback<React.ChangeEventHandler<HTMLInputElement>>(event => {
+        setDraft(event.target.checked)
+    }, [])
+
+    return (
+        <Dialog
+            className="modal-body modal-body--top-third p-4 rounded border"
+            onDismiss={onCancel}
+            aria-labelledby={MODAL_LABEL_ID}
+        >
+            <h3 id={MODAL_LABEL_ID}>Publish changesets</h3>
+            <p className="mb-4">
+                Are you sure you want to attempt to publish all the selected changesets to the code hosts?
```

Comment:
My vote for removing it. In combination with "Are you sure" it sounds too defensive and would make me as a user think "wait, did I miss something?" instead of triggering the "ah, a confirmation popup" part of the brain

```suggestion
                Are you sure you want to publish all the selected changesets to the code hosts?
```

Reply:
Sure, let's try it.